README.md
# Django-boto

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/qnub/django-boto?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Is an implementation of `Django` integration with [Amazon AWS](http://aws.amazon.com/)
services through [boto](https://github.com/boto/boto) module.

# Installation

To install:

    easy_install django-boto

or:

    pip install django-boto

# Configuration

## DEFAULT_FILE_STORAGE

Can't be used as default file storage system because of `path` not
implemented but you still can pass `S3Storage` object as storage
option to a `FileField`:

    from django.db import models
    from django_boto.s3.storage import S3Storage

    s3 = S3Storage()

    class Car(models.Model):
        ...
        photo = models.ImageField(storage=s3)

## Other settings.py options

`AWS_ACCESS_KEY_ID`
    *(required for default file storage use)* **Access Key ID** from
    **Security Credentials** settings on AWS service. Required for using
    as default storage.

`AWS_SECRET_ACCESS_KEY`
    *(required for default file storage use)* **Secret Access Key** from
    **Security Credentials** settings on AWS service. Required for using
    as default storage.

`AWS_ACL_POLICY`
    Default canned ACL for objects saved. Defaults to `public-read`.

`BOTO_S3_BUCKET`
    Amazon S3 `bucket` name. Default set to `AWS_ACCESS_KEY_ID`.

`BOTO_S3_HOST`
    Amazon S3 hostname. Default to `s3.amazonaws.com`.

`BOTO_BUCKET_LOCATION`
    Amazon datacenter location. Default to `US Classic Region`.

`AWS_S3_FORCE_HTTP_URL`
    Default to `False`. This settings allow you forcing S3 to return http links to files (if you have problem with SSL).

# Usage

## Manual S3Storage usage

If you need to use it manually - you can pass `bucket_name`
(as `BOTO_S3_BUCKET`), `key` (as `AWS_ACCESS_KEY_ID`),
`secret` (as `AWS_SECRET_ACCESS_KEY`) and `location`
(as `BOTO_BUCKET_LOCATION`) directly to storage constructor:

    from django_boto.s3.storage import S3Storage

    s3 = S3Storage(bucket_name='another-bucket', key='another-key',
        secret='another-secret', location='EU')


`S3Storage` is a typical [Django storage system](http://readthedocs.org/docs/django/en/1.4/ref/files/storage.html#the-storage-class), only `path`
is not implemented and `created_time` and `accessed_time` return
same value as `modified_time`.

## Upload shortcut

You can use a shortcut for simple uploads:

    from django_boto.s3 import upload

    upload(filename, name=None, prefix=False, bucket_name=False, key=None,
           secret=None, host=None, expires=0, query_auth=False, force_http=True,
           policy=None)

where:

`filename`:
    `string` filesystem path to file or **django** `File` instance
    or **python** `file` object instance;

`name`:
    `string` target Django's name for uploading the file;

`prefix`:
    `string` path prefix to filename in `s3.amazonaws.com` url. Like
    filename `/images/image.jpg` with `avatars` prefix convert to
    `avatars/image.jpg` in amazon url;

`bucket_name`:
    name of bucket (if not exists - system try to create it) in amazon
    S3;

`key`:
    `AWS_ACCESS_KEY_ID` replacement;

`secret`:
    `AWS_SECRET_ACCESS_KEY` replacement.

`host`:
    `BOTO_S3_HOST` replacement.

`expires`:
    `int` How long should private links be valid for.

`query_auth`:
    `bool` Should the url be generated with a valid signature?
    Required for private files.

`force_http`:
    `bool` Should the generated url be http?

`policy`:
    `string` Canned acl string for uploaded objects.

Last nine options are optional. If not set - it's taken from the `settings.py`
or defaults are used.

`upload()` returns a generated URL for a file download.


CHANGELOG.md
### 0.3.12 (12.07.2017)

* optional ACL dependence (thanx @gebetix)

### 0.3.11 (07.05.2017)

* add manifest

### 0.3.10 (20.10.2014)

* django 1.10 compatibilty (thanx @albertvaka)

### 0.3.8 (20.10.2014)

* django 1.7 compatibilty (thanx @buff3r)

### 0.3.7 (20.10.2014)

* RADME fixes (thanx @radeksimko)

### 0.3.5 (19.09.2014)

* add manifest

### 0.3.3, 0.3.4 (19.09.2014)

* setup fixes

### 0.3.2(09.09.2014)

* python 3 compatibility (thanx @axik)
* Add AWS_S3_FORCE_HTTP_URL setting (thanx @axik)
* Fix typos in README. (thanx @axik)
* Fixed bug in django admin page see https://code.djangoproject.com/ticket/19538 (thanx @axik)


### 0.3.1 (25.03.2014)

* fix `S3Storage._open()` signature (thanks @buff3r)
* fix non latin download url (thanks @caullla)

### 0.3.0 (05.02.2014)

* handle S3ResponseError (thanks @andrewjanssen)

### 0.2.3 (09.12.2013)

* changed default added in 0.2.2 for generated urls to use https rather than http (thanks @stuart-warren)
* allow creating write-once files, by setting replace=False on upload (thanks @stuart-warren)
* override Storage.get_available_name to stop creating extra files when you intended to replace existing file. (thanks @stuart-warren)
* by default will now replace existing files rather than creating a new file with suffix (thanks @stuart-warren)
* tidy up duplicate code in shortcuts (thanks @stuart-warren)
* reduce the default expire period to 30 seconds (thanks @stuart-warren)


### 0.2.2 (08.12.2013)

* allow using s3 based storage not hosted by Amazon (thanks @stuart-warren)
* added more shortcut functions (delete, get_url, download) (thanks @stuart-warren)
* allow to set permissions on bucket/key other than 'public-read' (thanks @stuart-warren)
* fixed S3Storage.url (thanks @stuart-warren)
* hopefully all backwards compatible (thanks @stuart-warren)

### 0.2.1 (03.08.2013)

* fixed file saving bug (thanks @loweroctave)

### 0.2 (02.08.2013)

* Upgraded upload shortcut to use custom filename (thanks @ivan-yurin)

### 0.1.6b (23.04.2012)


* fixed returned url

### 0.1.5b (23.04.2012)


* key uploading exception wrapped to IOError exception and added message

### 0.1.4b (21.04.2012)


* Seek(0) saved file before save

### 0.1.3b (21.04.2012)


* Fixed S3Storage._save
* Fixed S3Storage.__init__
* Fixed upload shortcut

### 0.1.2b (21.04.2012)


* Fixed S3Storage._save

### 0.1.1b (18.04.2012)


* Fixed S3Storage.listdir
* Fixed README.

### 0.1b (18.04.2012)

* S3 backend for django file storage


django_boto/tests.py
# -*- coding: utf-8 -*-

import string
import random
import logging
import urllib2
from os import path

from django.test import TestCase
from django.core.files.base import ContentFile

from s3 import upload
from s3.storage import S3Storage
from settings import BOTO_S3_BUCKET

logger = logging.getLogger(__name__)

local_path = path.realpath(path.dirname(__file__))


def get_string(lngth):
    strn = ''

    for i in xrange(lngth):
        strn += random.choice(string.letters)

    return strn


class BotoTest(TestCase):
    """
    Testing Amazon S3.
    """

    def test_storage(self):
        """
        Storage testing.
        """
        text = ''
        storage = S3Storage(host='s3.amazonaws.com')

        file_length = random.randrange(300, 1300)
        text = get_string(file_length)

        filename_length = random.randrange(5, 12)
        filename = get_string(filename_length)

        self.assertFalse(storage.exists(filename))

        test_file = ContentFile(text)
        test_file.name = filename
        uploaded_url = upload(test_file, host='s3.amazonaws.com')

        self.assertTrue(storage.exists(filename))

        url = 'http://' + BOTO_S3_BUCKET + '.s3.amazonaws.com/' + filename
        self.assertEqual(uploaded_url, url)

        page = urllib2.urlopen(uploaded_url)

        self.assertEqual(text, page.read())
        self.assertEqual(len(text), storage.size(filename))
        self.assertEqual(url, storage.url(filename))

        storage.delete(filename)
        self.assertFalse(storage.exists(filename))


django_boto/s3/shortcuts.py
# -*- coding: utf-8 -*-

from django.core.files import File
from .storage import S3Storage


def upload(filename, name=None, prefix=False, bucket_name=False, key=None,
           secret=None, host=None, expires=30, query_auth=False, force_http=False,
           policy=None, replace=True):
    """
    Uploading files to Amamzon S3.
    Returns String.
    """
    if isinstance(filename, str):
        fl = open(filename, 'rb')
    elif isinstance(filename, File):
        fl = filename
    else:
        raise TypeError('File must be file or string instance.')

    if not name:
        name = fl.name

    full_path = _get_name(name, prefix)

    s3 = S3Storage(bucket_name=bucket_name, key=key, secret=secret, host=host,
                   policy=policy, replace=replace)
    s3.save(full_path, fl)

    return s3.url(full_path, expires, query_auth, force_http)


def get_url(name=None, prefix=False, bucket_name=False, key=None,
            secret=None, host=None, expires=30, query_auth=False, force_http=False):
    """
    Get Url for key on Amazon S3.
    Returns String.
    """
    full_path = _get_name(name, prefix)

    s3 = S3Storage(bucket_name=bucket_name, key=key, secret=secret, host=host)

    return s3.url(full_path, expires, query_auth, force_http)


def download(name=None, prefix=False, bucket_name=False, key=None,
             secret=None, host=None):
    """
    Download file from Amazon S3.
    Returns TemporaryFile().
    """
    full_path = _get_name(name, prefix)

    s3 = S3Storage(bucket_name=bucket_name, key=key, secret=secret, host=host)

    return s3.open(full_path)


def remove(name=None, prefix=False, bucket_name=False, key=None,
           secret=None, host=None):
    """
    Deletes file from Amazon S3.
    """
    full_path = _get_name(name, prefix)

    s3 = S3Storage(bucket_name=bucket_name, key=key, secret=secret, host=host)

    s3.delete(full_path)


def _get_name(name, prefix):

    if prefix:
        if prefix.endswith('/'):
            full_path = prefix + name
        else:
            full_path = prefix + '/' + name
    else:
        full_path = name

    return full_path


django_boto/s3/storage.py
# -*- coding: utf-8 -*-
import logging
from tempfile import TemporaryFile

from dateutil.parser import parse as parse_date
from django.core.files.storage import Storage
from django.utils import timezone

try:
    from django.utils.deconstruct import deconstructible
except ImportError:
    def deconstructible(func):
        return func

from django.conf import settings as _settings
from boto import connect_s3
from boto.s3.connection import Location
from boto.exception import S3CreateError, S3ResponseError
from django_boto import settings


logger = logging.getLogger(__name__)


@deconstructible
class S3Storage(Storage):

    """
    Storage class.
    """

    def __init__(self, bucket_name=None, key=None, secret=None, location=None,
                 host=None, policy=None, replace=True, force_http_url=False):

        self.bucket_name = bucket_name if bucket_name else settings.BOTO_S3_BUCKET
        self.key = key if key else settings.AWS_ACCESS_KEY_ID
        self.secret = secret if secret else settings.AWS_SECRET_ACCESS_KEY
        self.location = location if location else settings.BOTO_BUCKET_LOCATION
        self.host = host if host else settings.BOTO_S3_HOST
        self.policy = policy if policy else settings.AWS_ACL_POLICY
        self.force_http = force_http_url if force_http_url else settings.AWS_S3_FORCE_HTTP_URL
        self.replace = replace

        if hasattr(Location, self.location):
            self.location = getattr(Location, self.location)

        self._bucket = None

    @property
    def bucket(self):
        if not self._bucket:
            self.s3 = connect_s3(
                aws_access_key_id=self.key,
                aws_secret_access_key=self.secret,
                host=self.host)
            try:
                self._bucket = self.s3.create_bucket(
                    self.bucket_name, location=self.location, policy=self.policy)
            except (S3CreateError, S3ResponseError):
                self._bucket = self.s3.get_bucket(self.bucket_name)
        return self._bucket

    def delete(self, name):
        """
        Delete file.
        """
        self.bucket.new_key(name).delete()

    def exists(self, name):
        """
        Existing check.
        """
        return self.bucket.new_key(name).exists()

    def _list(self, path):
        result_list = self.bucket.list(path, '/')

        for key in result_list:
            yield key.name

    def listdir(self, path):
        """
        Catalog file list.
        """
        return [], self._list(path)

    def size(self, name):
        """
        File size.
        """
        return self.bucket.lookup(name).size

    def url(self, name, expires=30, query_auth=False, force_http=None):
        """
        URL for file downloading.
        """
        if not force_http:
            force_http = self.force_http
        if name == 'admin/':
            # https://code.djangoproject.com/ticket/19538
            return _settings.S3_URL + 'admin/'
        key = self.bucket.get_key(name)
        return key.generate_url(expires, query_auth=query_auth,
                                force_http=force_http)

    def _open(self, name, mode='rb'):
        """
        Open file.
        """
        result = TemporaryFile()
        self.bucket.get_key(name).get_file(result)

        return result

    def _save(self, name, content):
        """
        Save file.
        """
        key = self.bucket.new_key(name)
        content.seek(0)

        if self.replace:
            try:
                key.set_contents_from_file(content, replace=True)
            except Exception as e:
                raise IOError('Error during uploading file - %s' % e)
        else:
            if key.exists():
                raise IOError(
                    'File already exists and can\'t be replaced - %s' % name)
            else:
                try:
                    key.set_contents_from_file(content, replace=False)
                except Exception as e:
                    raise IOError('Error during uploading file - %s' % e)

        content.seek(0, 2)
        orig_size = content.tell()
        saved_size = key.size

        if saved_size == orig_size:
            if self.policy:
                key.set_acl(self.policy)
        else:
            key.delete()

            raise IOError('Error during saving file %s - saved %s of %s bytes'
                          % (name, saved_size, orig_size))

        return name

    def get_available_name(self, name, max_length=None):
        """
        Returns a filename that's free on the target storage system, and
        available for new content to be written to.
        Handled by Boto itself.
        """

        return name

    def modified_time(self, name):
        """
        Last modification time.
        """
        return timezone.make_naive(
            parse_date(self.bucket.lookup(name).last_modified),
            timezone.get_default_timezone())

    created_time = accessed_time = modified_time
